runtime/src/qio/qio.c

```diff
@@ -2879,38 +2886,113 @@ qioerr _qio_buffered_write(qio_channel_t* ch, const void* ptr, ssize_t len, ssiz
+      err = QIO_EEOF;
+      goto error;
     }
 
-    // make sure we have buffer space. (require calls advance_cached)
-    err = _qio_channel_require_unlocked(ch, toWrite, 1);
-    eof = 0;
-    if( qio_err_to_int(err) == EEOF ) eof = 1;
-    else if( err ) goto error;
-
-    // figure out the end of the data to copy
-    start = _right_mark_start_iter(ch);
-    end = start;
-    gotlen = qbuffer_iter_num_bytes_after(&ch->buf, end);
-    if( toWrite < gotlen ) gotlen = toWrite;
-    qbuffer_iter_advance(&ch->buf, &end, gotlen);
-
-    // now copy the data in to the buffer.
-    err = qbuffer_copyin(&ch->buf, start, end, (char *) ptr + (len-remaining),
-                         gotlen);
+    // flush the write-behind portion of the buffer before attempting to write more
+    err = _qio_channel_flush_qio_unlocked(ch);
     if( err ) goto error;
 
-    // now move start forward.
-    _set_right_mark_start(ch, end.offset);
+    // trim any excess bytes from the current buffer part
+    int64_t trim_bytes = qbuffer_end_offset(&ch->buf) -
+                         qbuffer_start_offset(&ch->buf);
+    qbuffer_trim_back(&ch->buf, trim_bytes);
 
-    // did we get to a different part? If so, call write()
-    err = _qio_buffered_behind(ch, false);
-    if( err ) goto error;
-    remaining -= gotlen;
+    // write with a direct system call (may require multiple iterations for write/pwrite)
+    while ( remaining > 0 ) {
+      num_written = 0;
+      switch (method) {
+        case QIO_METHOD_READWRITE:
+          err = qio_int_to_err(sys_write(ch->file->fd, ptr, remaining, &num_written));
+          break;
+        case QIO_METHOD_PREADPWRITE:
+          err = qio_int_to_err(sys_pwrite(ch->file->fd, ptr, remaining, _right_mark_start(ch), &num_written));
+          break;
+        case QIO_METHOD_FREADFWRITE:
+          if( ch->file->fp ) {
+            num_written_u = fwrite(ptr, 1, remaining, ch->file->fp);
+            err = 0;
+            if( num_written_u == 0 ) {
+              err = qio_int_to_err(ferror(ch->file->fp));
+            }
+            num_written = num_written_u;
+          } else {
+            QIO_GET_CONSTANT_ERROR(err, EINVAL, "missing file pointer");
+            num_written = 0;
+          }
+          break;
+          case QIO_METHOD_MMAP:
+          break;
+          case QIO_METHOD_MEMORY:
+          break;
+      }
+      if( err ) {
+        *amt_written = num_written + len - remaining;
+        goto error;
+      }
+      // move the ptr and right_mark_start forward by 'num_written' bytes
+      ptr = qio_ptr_add((void*) ptr, num_written);
+      _add_right_mark_start(ch, num_written);
+      remaining -= num_written;
+    }
+
+    // shift the buffer position 'len' bytes to the right
+    int64_t new_start = qbuffer_end_offset(&ch->buf) + len;
+    qbuffer_reposition(&ch->buf, new_start);
+    ch->av_end = new_start;
+
+    // re-setup the buffer for any future buffered writes
+    _qio_buffered_setup_cached(ch);
```

Comment:
Perhaps it would be better not to do this here, and let the setup occur on a cached read? It is a performance question.

Reply:
I'll try moving it to the beginning of the buffered write code and see if that fixes the correctness issues I was seeing without this line here.